### PR TITLE
fix(device): throw during client init for invalid HTTP transport with…

### PR DESCRIPTION
**Pull request by drwill-ms**
For GitHub issue #1917

### CODE REVIEW

File: iothub/device/src/ClientFactory.cs [hunk 1]
```diff
@@ -28,7 +31,7 @@ internal class ClientFactory
         /// <param name="authenticationMethod">The authentication method that is used</param>
         /// <param name="options">The options that allow configuration of the device client instance during initialization.</param>
         /// <returns>InternalClient</returns>
-        public static InternalClient Create(string hostname, IAuthenticationMethod authenticationMethod, ClientOptions options = default)
+        internal static InternalClient Create(string hostname, IAuthenticationMethod authenticationMethod, ClientOptions options = default)
```
Comment (drwill-ms) [inline iothub/device/src/ClientFactory.cs:34]: Class is internal, and so these are effectively internal. I guess I finally got tired of the dissonance and misleading visibility spec.

File: iothub/device/src/ClientFactory.cs [hunk 2]
```diff
@@ -82,13 +98,19 @@ public static InternalClient Create(string hostname, string gatewayHostname, IAu
...
-                && authenticationMethod is DeviceAuthenticationWithX509Certificate
-                && ((DeviceAuthenticationWithX509Certificate)authenticationMethod).ChainCertificates != null)
+                && authenticationMethod is DeviceAuthenticationWithX509Certificate certificate
+                && certificate.ChainCertificates != null)
             {
                 throw new ArgumentException("Certificate chains are only supported on Amqp_Tcp_Only and Mqtt_Tcp_Only");
             }
 
-            IotHubConnectionStringBuilder connectionStringBuilder = IotHubConnectionStringBuilder.Create(hostname, gatewayHostname, authenticationMethod);
+            if (!string.IsNullOrWhiteSpace(options?.ModelId)
```
Comment (drwill-ms) [inline iothub/device/src/ClientFactory.cs:107]: One of 3 checks.

File: iothub/device/src/ClientFactory.cs [hunk 3]
```diff
@@ -160,6 +193,12 @@ public static InternalClient Create(string hostname, IAuthenticationMethod authe
                 throw new ArgumentNullException(nameof(authenticationMethod));
             }
 
+            if (!string.IsNullOrWhiteSpace(options?.ModelId)
```
Comment (drwill-ms) [inline iothub/device/src/ClientFactory.cs:196]: One of 3 checks.

File: iothub/device/src/ClientFactory.cs [hunk 4]
```diff
@@ -366,12 +419,19 @@ internal static ITransportSettings[] GetTransportSettings(TransportType transpor
                 throw new ArgumentOutOfRangeException(nameof(connectionString), "Must specify at least one TransportSettings instance");
             }
 
-            IotHubConnectionStringBuilder builder = IotHubConnectionStringBuilder.CreateWithIAuthenticationOverride(
+            if (!string.IsNullOrWhiteSpace(options?.ModelId)
```
Comment (drwill-ms) [inline iothub/device/src/ClientFactory.cs:422]: One of 3 checks.

File: iothub/device/tests/DeviceClientTests.cs [hunk 5]
```diff
@@ -1684,6 +1685,94 @@ public void DeviceClient_CreateFromAuthenticationMethod_SasTokenTimeToLiveRenewa
             authMethod.RefreshesOn.Should().BeCloseTo(expectedRefreshTimeFromSdkDefault, (int)buffer.TotalMilliseconds);
         }
 
+        [TestMethod]
```
Comment (drwill-ms) [inline iothub/device/tests/DeviceClientTests.cs:1688]: Unit tests to validate it throws when it should, and doesn't throw when it shouldn't. Any gaps?

File: iothub/device/src/ClientFactory.cs [hunk 6]
```diff
@@ -82,13 +98,19 @@ public static InternalClient Create(string hostname, string gatewayHostname, IAu
...
+                && certificate.ChainCertificates != null)
             {
                 throw new ArgumentException("Certificate chains are only supported on Amqp_Tcp_Only and Mqtt_Tcp_Only");
             }
 
-            IotHubConnectionStringBuilder connectionStringBuilder = IotHubConnectionStringBuilder.Create(hostname, gatewayHostname, authenticationMethod);
+            if (!string.IsNullOrWhiteSpace(options?.ModelId)
+                && transportType == TransportType.Http1)
+            {
+                throw new InvalidOperationException("PnP is not supported over the HTTP transport.");
```
Comment (timtay-microsoft) [inline iothub/device/src/ClientFactory.cs:110]: Abbreviations may cause confusion here. How about "Plug and play is not supported over HTTP"?

File: iothub/device/src/ClientFactory.cs [hunk 7]
```diff
@@ -160,6 +193,12 @@ public static InternalClient Create(string hostname, IAuthenticationMethod authe
                 throw new ArgumentNullException(nameof(authenticationMethod));
             }
 
+            if (!string.IsNullOrWhiteSpace(options?.ModelId)
+                && transportSettings.Any(x => x.GetTransportType() == TransportType.Http1))
+            {
+                throw new InvalidOperationException("PnP is not supported over the HTTP transport.");
```
Comment (timtay-microsoft) [inline iothub/device/src/ClientFactory.cs:199]: Same comment here

**Review by timtay-microsoft — APPROVED** [review]
One small comment, otherwise looks good!

**Review by azabbasi — APPROVED** [review]

File: iothub/device/src/ClientFactory.cs (hunk 6 above)
Comment (drwill-ms) [reply iothub/device/src/ClientFactory.cs:110]: Changed

**Pull request merged**